src/main/kotlin/org/rust/lang/core/psi/ext/RsDocAndAttributeOwner.kt

```diff
@@ -79,24 +83,120 @@ inline val RsDocAndAttributeOwner.attributeStub: RsAttributeOwnerStub?
  * Returns [QueryAttributes] for given PSI element.
  */
 val RsDocAndAttributeOwner.queryAttributes: QueryAttributes
+    get() = getQueryAttributes(null)
+
+/** [explicitCrate] is passed to avoid resolve triggering */
+fun RsDocAndAttributeOwner.getQueryAttributes(
+    explicitCrate: Crate?,
+    stub: RsAttributeOwnerStub? = attributeStub
+): QueryAttributes {
+    testAssert { !DumbService.isDumb(project) }
+    return when {
+        stub?.hasAttrs == false -> QueryAttributes.EMPTY
+        stub?.hasCfgAttr == false -> rawAttributesUnchecked
+        else -> getQueryAttributesNoStub(explicitCrate)
+    }
+}
+
+/** [explicitCrate] is passed to avoid resolve triggering */
+private fun RsDocAndAttributeOwner.getQueryAttributesNoStub(explicitCrate: Crate?): QueryAttributes {
+    if (!CFG_ATTRIBUTES_ENABLED_KEY.asBoolean()) return rawAttributesUnchecked
+    val crate = explicitCrate ?: containingCrate ?: return rawAttributesUnchecked
+    val evaluator = newCfgEvaluator(crate)
+    return QueryAttributes(evaluator.expandCfgAttrs(rawMetaItems))
+}
+
+private fun CfgEvaluator.expandCfgAttrs(rawMetaItems: Sequence<RsMetaItem>): Sequence<RsMetaItem> {
+    return rawMetaItems.flatMap {
+        if (it.name == "cfg_attr") {
+            val list = it.metaItemArgs?.metaItemList?.iterator() ?: return@flatMap emptySequence()
+            val condition = list.nextOrNull() ?: return@flatMap emptySequence()
+            if (evaluateCondition(condition) != ThreeValuedLogic.False) {
+                expandCfgAttrs(list.asSequence())
+            } else {
+                emptySequence()
+            }
+        } else {
+            sequenceOf(it)
+        }
+    }
+}
+
+fun RsDocAndAttributeOwner.getTraversedRawAttributes(withCfgAttr: Boolean): QueryAttributes {
```

Comment:
[minor] `cfgAttr` name can be treated as "cfg attribute", not "cfg_attr attribute". Maybe use `withCfgAttrAttribute` name or even something else?